File: C_plus_plus/Add_Binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int la = a.size() , lb = b.size();
        int lenth = la > lb ? lb : la;
        int carry = 0;
        string result;
        for(int i = 1 ; i <= lenth ; ++i){
            result.insert(result.begin(),(a[la-i]-'0'+b[lb-i]-'0'+carry)%2+'0');
            carry = (a[la-i]-'0'+b[lb-i]-'0'+carry)/2;
        }
        if(la>lb){
            for(int i = lenth + 1 ; i <= la ; ++i){
                result.insert(result.begin(),(a[la-i]-+carry)%2+'0');
                carry = (a[la-i]-'0'+carry)/2;                
            }
        }
        if(la<lb){
            for(int i = lenth + 1 ; i <= lb ; ++i){
                result.insert(result.begin(),(b[lb-i]-+carry)%2+'0');
                carry = (b[lb-i]-'0'+carry)/2;                
            }
        }
        if(carry) result.insert(result.begin(),carry+'0');
        return result;
    }
};

File: java/Add_Binary.java
// Time and space is both O(n)

public class Solution {
    public String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int i = a.length()-1, j = b.length()-1;
        int carry = 0, sum;
        while (i >= 0 || j >= 0) {
            if (i < 0) {
                sum = carry + b.charAt(j--)-'0';
            }
            else if (j < 0) {
                sum = carry + a.charAt(i--)-'0';
            }
            else {
                sum = carry + a.charAt(i--)-'0' + b.charAt(j--)-'0';
            }
            sb.insert(0, (char)(sum % 2 + '0'));
            carry = sum / 2;
        }
        if (carry > 0)
            sb.insert(0,'1');
        return sb.toString();
    }
}

// It's interesting to simplify the code like this :

public class Solution {
    public String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int i = a.length()-1, j = b.length()-1;
        int carry = 0;
        while (i >= 0 || j >= 0 || carry > 0) {
            int sum = carry + 
                  (i < 0 ? 0 : a.charAt(i--)-'0') + 
                  (j < 0 ? 0 : b.charAt(j--)-'0');
            sb.insert(0, (char)(sum % 2 + '0'));
            carry = sum / 2;
        }
        return sb.toString();
    }
}
